src/server/opc_tcp_async.cpp

```diff
@@ -301,28 +312,54 @@ namespace
   void OpcTcpServer::Listen()
   {
     std::clog << "opc_tcp_async| Running server." << std::endl;
+
+    std::cout << "opc_tcp_async| Waiting for client connection at: " << acceptor.local_endpoint().address() << ":" << acceptor.local_endpoint().port() <<  std::endl;
+    acceptor.listen();
+
     Accept();
   }
 
   void OpcTcpServer::Shutdown()
   {
     std::clog << "opc_tcp_async| Shutting down server." << std::endl;
-    Clients.clear();
     acceptor.close();
+
+    // Actively shutdown OpcTcpConnections to clear open async requests from worker
+    // thread.
+    // Warning: the Clients container may be modified by OpcTcpConnections::GoodBye
+    // when calling Stop() which makes our iterator invalid. So have a copy of thiss
+    // container to iterate have a stable iterator.
+    Mutex.lock();
+    std::set<std::shared_ptr<OpcTcpConnection>> tmp(Clients);
+    Mutex.unlock();
+    auto i = tmp.begin();
+    while (i != tmp.end()) {
+      (*i)->Stop();
+      ++i;
+    }
+
+    // clear possibly remaining Client's
+
+    { std::unique_lock<std::mutex> lock(Mutex);
+      Clients.clear();
+    }
   }
```

Comment:
As OpcTcpServer::Clients is accessed from different threads all operations on it must be guarded by a lock.

The loop to stop all clients on the other hand must not lock Clients, as OpcTcpConnection::Stop() interrupts all pending read/write requests, which may then call GoodBye() and try to deregister themselves from the Clients container. OpcTcpConnection::Stop() waits for completion of all pending operations (therefore the boost::promise) to make sure these operations do not use an already deleted OpcTcpConnection or OpcTcpServer instance. So if we have a single locked block for copy/stop loop/clear we run into a deadlock, because the pending operations will not return until OpcTcpServer::Clients is unlocked.